logger/logger.go

```diff
@@ -0,0 +1,113 @@
+package logger
+
+import (
+	"sync"
+	"time"
+)
+
+// LogLine is the structure used to hold a log line
+type LogLine struct {
+	Message   string
+	LogLevel  LogLevel
+	Args      []interface{}
+	Timestamp time.Time
+}
+
+func newLogLine(message string, logLevel LogLevel, args ...interface{}) *LogLine {
+	return &LogLine{
+		Message:   message,
+		LogLevel:  logLevel,
+		Args:      args,
+		Timestamp: time.Now(),
+	}
+}
+
+// logger is the primary structure used to interact with the productive code
+type logger struct {
+	name      string
+	mutLevel  sync.RWMutex
+	logLevel  LogLevel
+	logOutput LogOutputHandler
+}
+
+// newLogger create a new logger instance
+func newLogger(name string, logLevel LogLevel, logOutput LogOutputHandler) *logger {
+	log := &logger{
+		name:      name,
+		logLevel:  logLevel,
+		logOutput: logOutput,
+	}
+
+	return log
+}
+
+func (l *logger) shouldOutput(compareLogLevel LogLevel) bool {
+	l.mutLevel.RLock()
+	shouldOutput := l.logLevel > compareLogLevel
+	l.mutLevel.RUnlock()
+
+	return shouldOutput
+}
+
+// Trace outputs a tracing log message with optional provided arguments
+func (l *logger) Trace(message string, args ...interface{}) {
+	if l.shouldOutput(LogTrace) {
```

Comment:
this function and the next 3 are exactly the same beside of the log level. Maybe you can add a function a call that function from all these 4:
```GOLANG
func (l *logger) outputMessageFromLogLevel(level LogLevel, message string, args ...interface{}) {
   if l.shouldOutput(level) {
		return
   }

   logLine := newLogLine(message, level, args...)
   l.logOutput.Output(logLine)
}

Reply:
Done